pilot/cmd/pilot-discovery/main.go

```diff
@@ -113,64 +98,78 @@ func hasKubeRegistry() bool {
-		"Initial connection window size for MCP's gRPC connection")
-
-	// Config Controller options
-	discoveryCmd.PersistentFlags().BoolVar(&serverArgs.Config.DisableInstallCRDs, "disable-install-crds", true,
-		"Disable discovery service from verifying the existence of CRDs at startup and then installing if not detected.  "+
-			"It is recommended to be disable for highly available setups.")
-	_ = discoveryCmd.PersistentFlags().MarkDeprecated("disable-install-crds",
-		"Setting this flag has no effect. Install CRD definitions directly or with the operator")
-	discoveryCmd.PersistentFlags().StringVar(&serverArgs.Config.FileDir, "configDir", "",
-		"Directory to watch for updates to config yaml files. If specified, the files will be used as the source of config, rather than a CRD client.")
-	discoveryCmd.PersistentFlags().StringVarP(&serverArgs.Config.ControllerOptions.WatchedNamespace, "appNamespace",
-		"a", metav1.NamespaceAll,
-		"Restrict the applications namespace the controller manages; if not set, controller watches all namespaces")
-	discoveryCmd.PersistentFlags().DurationVar(&serverArgs.Config.ControllerOptions.ResyncPeriod, "resync", 60*time.Second,
-		"Controller resync interval")
-	discoveryCmd.PersistentFlags().StringVar(&serverArgs.Config.ControllerOptions.DomainSuffix, "domain", "cluster.local",
-		"DNS domain suffix")
-	discoveryCmd.PersistentFlags().StringVar(&serverArgs.Config.ControllerOptions.TrustDomain, "trust-domain", "",
-		"The domain serves to identify the system with spiffe")
-	discoveryCmd.PersistentFlags().StringVar(&serverArgs.Service.Consul.ServerURL, "consulserverURL", "",
-		"URL for the Consul server")
-
-	// using address, so it can be configured as localhost:.. (possibly UDS in future)
-	discoveryCmd.PersistentFlags().StringVar(&serverArgs.DiscoveryOptions.HTTPAddr, "httpAddr", ":8080",
-		"Discovery service HTTP address")
-	discoveryCmd.PersistentFlags().StringVar(&serverArgs.DiscoveryOptions.HTTPSAddr, "httpsAddr", ":15017",
-		"Injection and validation service HTTPS address")
-	discoveryCmd.PersistentFlags().StringVar(&serverArgs.DiscoveryOptions.GrpcAddr, "grpcAddr", ":15010",
-		"Discovery service grpc address")
-	discoveryCmd.PersistentFlags().StringVar(&serverArgs.DiscoveryOptions.SecureGrpcAddr, "secureGrpcAddr", ":15012",
-		"Discovery service grpc address, with https")
-	discoveryCmd.PersistentFlags().StringVar(&serverArgs.DiscoveryOptions.MonitoringAddr, "monitoringAddr", ":15014",
-		"HTTP address to use for pilot's self-monitoring information")
-	discoveryCmd.PersistentFlags().BoolVar(&serverArgs.DiscoveryOptions.EnableProfiling, "profile", true,
-		"Enable profiling via web interface host:port/debug/pprof")
+
+	serverArgs = bootstrap.NewPilotArgs(func(p *bootstrap.PilotArgs) {
+		// Process commandline args.
+		discoveryCmd.PersistentFlags().StringSliceVar(&p.Service.Registries, "registries",
```

Comment:
Why are CLI flags initialized here? I think the only requirement is that this flags are created after defaults are applied. Calling NewPilotArgs() in the same init() block before line 116 in the original file should satisfy that requirement.

Reply:
Yes,. That is possible. Changed PTAL